### PR TITLE
Accessibility

### PR DESCRIPTION
**Pull request by Wooble**
This PR fixes some minor accessibility issues a scan of our catalog found in the bootstrap3 theme.

TODO:
- [x] Add missing labels to a few form elements
- [ ] Fix places where non-unique <title> elements appear (e.g.: "Help", "Search Results" for searches with no terms, only facets)
- [ ] "Clear" and "Set" buttons in advanced search have same label on duplicates. (I'm not sure this is an actual problem, since the duplicate buttons do literally the same thing, but it's getting flagged as violating WCAG 2.0 AA 2.4.6 by SortSite)

### CODE REVIEW

File: themes/bootstrap3/templates/RecordTab/similaritemscarousel.phtml [hunk 1]
```diff
@@ -41,10 +41,10 @@
     </div>
 
     <!-- Controls -->
-    <a class="left carousel-control" href="#similar-items-carousel" role="button" data-slide="prev">
+    <a class="left carousel-control" href="#similar-items-carousel" role="button" data-slide="prev" aria-label="Previous">
```
Comment (crhallberg) [inline themes/bootstrap3/templates/RecordTab/similaritemscarousel.phtml]: We have to make sure the labels are properly translated.

File: themes/bootstrap3/templates/eds/advanced.phtml [hunk 2]
```diff
@@ -22,9 +22,9 @@
           <option value="OR"><?=$this->transEsc("OR")?></option>
           <option value="NOT"><?=$this->transEsc("NOT")?></option>
         </select>
-        <input id="search_lookfor0_<?=$search ?>" name="lookfor0[]" class="adv-term-input form-control" type="text" value=""/>
+        <input id="search_lookfor0_<?=$search ?>" name="lookfor0[]" class="adv-term-input form-control" type="text" value="" aria-label="Search text <?=$search ?>"/>
         <span class="help-block hidden-xs"><?=$this->transEsc("in")?></span>
-        <select id="search_type0_<?=$search ?>" name="type0[]" class="adv-term-type form-control">
+        <select id="search_type0_<?=$search ?>" name="type0[]" class="adv-term-type form-control" aria-label="Search type <?=$search ?>">
```
Comment (crhallberg) [inline themes/bootstrap3/templates/eds/advanced.phtml]: You can concatenate translations for phrases like this for now but they should be tokenized in the future.

File: languages/en.ini [hunk 3]
```diff
@@ -773,6 +773,7 @@ Preferred Library = "Preferred Library"
 Prev = "Prev"
 Preview = "Preview"
 Preview from = "Preview from"
+Previous = "Previous"
```
Comment (demiankatz) [inline languages/en.ini]: Do we need a separate Previous translation when we already have Prev? If using an unabbreviated string is preferable for accessibility, that's fine with me -- I just want to be sure we're doing this intentionally.

File: languages/en.ini [hunk 4]
```diff
@@ -888,6 +889,7 @@ Search Mode = "Search Mode"
 Search Options = "Search Options"
 Search Results = "Search Results"
 search results of = "search results of"
+Search text = "Search text"
```
Comment (demiankatz) [inline languages/en.ini]: Would "Search terms" be better here? That seems more like a noun than a verb.

File: themes/bootstrap3/templates/search/advanced/build_page.phtml [hunk 5]
```diff
@@ -4,10 +4,12 @@ $(document).ready(function() {
   $('#groupPlaceHolder').removeClass('hidden');
   $('#new_search_template').addClass('hidden').detach().appendTo('[role="main"]');
   $('#new_search_template').find('.adv-search').removeAttr('id');
+  $('#new_search_template').find('.form-control').removeAttr('id');
```
Comment (demiankatz) [inline themes/bootstrap3/templates/search/advanced/build_page.phtml]: What exactly is going on here? I understand that we have an HTML template set up that we're copying to create the groups on the advanced search screen, and thus we need to remove the IDs in order to avoid HTML validation problems caused by duplicate IDs... and at least some of the IDs are there in the first place to allow labels to be associated with appropriate controls. But if we delete the IDs, aren't we breaking the labels? Has this whole thing been thought through all the way? @crhallberg, have you already looked into this? Do we need a better approach here to achieve labeling without risking duplicate IDs?

File: themes/bootstrap3/templates/search/facet-list.phtml [hunk 6]
```diff
@@ -11,6 +11,7 @@
     $searchAction .= urlencode($this->baseUriExtra);
     $urlBase .= '&amp;baseUriExtra=' . urlencode($this->baseUriExtra);
   }
+  $this->headTitle($this->translate('Facet List for ' . $this->facet));
```
Comment (demiankatz) [inline themes/bootstrap3/templates/search/facet-list.phtml]: It's not a good idea to create translation strings that rely on concatenation, because some languages use different word orders. It would be better to use a token, like `$this->translate('Facet List for %%field%%', ['%%field%%' => $this->facet])`. If we do that, we also need to add the new string to the language files.

File: themes/bootstrap3/templates/search/advanced/build_page.phtml (hunk 5 above)
Comment (Wooble) [reply themes/bootstrap3/templates/search/advanced/build_page.phtml]: The JS that copies the template onto the new search box also sets the appropriate IDs everywhere they were removed.

I think vufind/module/VuFind/tests/integration-tests/src/VuFindTest/Mink/AdvancedSearchTest.php at line 127 would fail if removing this ID was a regression.

Comment (crhallberg) [reply themes/bootstrap3/templates/search/advanced/build_page.phtml]: I saw this and saw no problems with it. Maybe we should remove the ID from the HTML to make this a little simpler. https://github.com/vufind-org/vufind/blob/23c75f5ef9477b3f921e9dbc58d07602018123c7/themes/bootstrap3/templates/search/advanced/layout.phtml#L108

Comment (demiankatz) [reply themes/bootstrap3/templates/search/advanced/build_page.phtml]: Reviewing the template brings back some memory of this. We shouldn't remove the id because this is serving two purposes -- it's not only the template html but also the non-Javascript fallback logic. That's probably something we should consider changing -- it's overcomplicated, for one thing -- but as it stands, if we took the id out without changing anything else, it would result in invalid labels for non-JS users.

File: themes/bootstrap3/templates/search/advanced/layout.phtml [hunk 7]
```diff
@@ -105,8 +105,8 @@
                     <div id="new_search_template">
                   <?php endif; ?>
                   <div id="search<?=$group . '_' . $search ?>" class="adv-search">
-                    <input name="lookfor<?=$group ?>[]" id="search_lookfor<?=$group . '_' . $search ?>" class="adv-term-input form-control" type="text"<?php if (isset($setQueries[$group][$search])): ?> value="<?=$this->escapeHtml($setQueries[$group][$search]->getString())?>"<?php endif; ?>>
-                    <select class="adv-term-type form-control" name="type<?=$group ?>[]">
+                    <input name="lookfor<?=$group ?>[]" id="search_lookfor<?=$group . '_' . $search ?>" class="adv-term-input form-control" type="text"<?php if (isset($setQueries[$group][$search])): ?> value="<?=$this->escapeHtml($setQueries[$group][$search]->getString())?>"<?php endif; ?> aria-label="<?=$this->transEsc("search_terms")?> <?=$search ?>">
```
Comment (demiankatz) [inline themes/bootstrap3/templates/search/advanced/layout.phtml]: Is including the $search number here actually necessary? Right now, it's not working -- because the Javascript works by duplicating elements from a template, when you add lines to the search group, they all get the same label of "Search type 0" / "Search terms 0". I think it would be much simpler to eliminate the number if we can get away with that. If we absolutely have to have non-unique labels, it will take some more work to get it working correctly and to use more appropriate translation strings with placeholders (instead of concatenating numbers onto labels).

**Comment by demiankatz** [comment]
@Wooble, I've just cherry-picked the "remove IDs from template" commit into the release-5.0 branch so we can fix that in 5.0.1. All integration tests are still passing, as you suggested they would. I've merged that back into this PR, reducing the number of changed files by 1.

I'll wait for more input from your end on some of the outstanding issues in the other comments. Let me know if you need anything from me in the meantime!

**Comment by Wooble** [comment]
Thanks; I'm not sure I'll be able to get to this before I return from vacation on the 27th

**Comment by demiankatz** [comment]
@Wooble, no rush at all -- enjoy your vacation!

File: themes/bootstrap3/templates/search/facet-list.phtml (hunk 6 above)
Comment (Wooble) [reply themes/bootstrap3/templates/search/facet-list.phtml]: I've fixed this, for English. It's not clear to me if there's an easy way to add things like this to the other translations using some sort of automated translating.

Comment (demiankatz) [reply themes/bootstrap3/templates/search/facet-list.phtml]: @Wooble, that's fine. There is no automated translating, but as long as new strings are added to the English file, I contact translators and ask for missing translations every time we plan a new release, so this will be included in that process.

File: themes/bootstrap3/templates/search/advanced/layout.phtml (hunk 7 above)
Comment (Wooble) [reply themes/bootstrap3/templates/search/advanced/layout.phtml]: Lighthouse seems fine with the numbers being dropped, so I've removed them for now instead of trying to hack together something that will actually make them unique and translated usefully.

**Comment by demiankatz** [comment]
Thanks, @Wooble, this all looks good, so I have gone ahead and merged to master. I see that you have some unresolved checkboxes here, so please feel free to open another PR if you feel more work is needed!

**Comment by Wooble** [comment]
@demiankatz Thanks. I think the second one probably fits better as a separate PR anyway since it's hard to test (as it's a problem across pages rather than a single page, so Lighthouse, for example, can't detect it at all), and the third one I'm not convinced is a real issue. IMO it doesn't matter much if you can confuse 2 buttons if they literally do the exact same thing.

**Comment by demiankatz** [comment]
@Wooble, that all makes sense to me. Regarding the duplicate buttons, is there a way to flag the second set as "screen reader ignore" to reduce the redundancy? Perhaps that would help, too.
